contracts/farm/LiquidityMining.sol

```diff
@@ -0,0 +1,567 @@
+		require(poolIdList[_poolToken] != 0, "Pool token not found");
+		address userAddress = _getUserAddress(_user);
+
+		uint256 poolId = _getPoolId(_poolToken);
+		PoolInfo storage pool = poolInfoList[poolId];
+		UserInfo storage user = userInfoMap[poolId][userAddress];
+
+		_updatePool(poolId);
+		_updateReward(pool, user);
+		_transferReward(user, userAddress, true);
+		_updateRewardDebt(pool, user);
+	}
+
+	/**
+	 * @notice withdraws pool tokens and transfers reward tokens
+	 * @param _poolToken the address of pool token
+	 * @param _amount the amount of pool tokens
+	 * @param _user the user address will be used to process a withdrawal (can be passed only by wrapper contract)
+	 */
+	function withdraw(
+		address _poolToken,
+		uint256 _amount,
+		address _user
+	) public {
+		require(poolIdList[_poolToken] != 0, "Pool token not found");
+		address userAddress = _getUserAddress(_user);
+
+		uint256 poolId = _getPoolId(_poolToken);
+		PoolInfo storage pool = poolInfoList[poolId];
+		UserInfo storage user = userInfoMap[poolId][userAddress];
+		require(user.amount >= _amount, "Not enough balance");
+
+		_updatePool(poolId);
+		_updateReward(pool, user);
+		_transferReward(user, userAddress, false);
+
+		user.amount = user.amount.sub(_amount);
+		pool.poolToken.safeTransfer(address(msg.sender), _amount); //sent to the user or wrapper
+		_updateRewardDebt(pool, user);
+		emit Withdraw(userAddress, _poolToken, _amount);
+	}
+
+	function _getUserAddress(address _user) internal view returns (address) {
+		address userAddress = msg.sender;
+		if (_user != address(0)) {
+			//only wrapper can pass _user parameter
+			require(msg.sender == wrapper && _user == tx.origin, "unauthorized");
+			userAddress = _user;
+		}
+		return userAddress;
+	}
+
+	function _updateReward(PoolInfo storage pool, UserInfo storage user) internal {
+		//update user accumulated reward
+		if (user.amount > 0) {
+			//add reward for the previous amount of deposited tokens
+			uint256 accumulatedReward = user.amount.mul(pool.accumulatedRewardPerShare).div(PRECISION).sub(user.rewardDebt);
+			user.accumulatedReward = user.accumulatedReward.add(accumulatedReward);
+		}
+	}
+
+	function _updateRewardDebt(PoolInfo storage pool, UserInfo storage user) internal {
+		//reward accumulated before amount update (should be subtracted during next reward calculation)
+		user.rewardDebt = user.amount.mul(pool.accumulatedRewardPerShare).div(PRECISION);
+	}
+
+	/**
+	 * @notice Send reward in SOV to the lockedSOV vault.
+	 * @param _user The user info, to get its reward share.
+	 * @param _userAddress The address of the user, to send SOV in its behalf.
+	 */
+	function _transferReward(
+		UserInfo storage _user,
+		address _userAddress,
+		bool _isClaimingReward
+	) internal {
+		uint256 userAccumulatedReward = _user.accumulatedReward;
+
+		/// @dev Transfer if enough SOV balance on this LM contract.
+		uint256 balance = SOV.balanceOf(address(this));
+		if (balance >= userAccumulatedReward) {
```

Comment:
If the balance is not enough this function will be successful but do nothing. This is a potentially dangerous behaviour.

Reply:
this is here, because the user should always be able to withdraw his pool tokens even if the contract cannot pay the reward it's owing the user. the user reward on the contract remains unchanged and the user can claim it later after the contract has been refilled.

however, it would be better to return a bool telling if the transfer was successful, so the calling function can decide what to do in such a case. then, the withdraw function can let it pass, but the claim function could decide to revert.

Reply:
We have `_isClaimingReward` flag, so an error will be thrown is case of an invokation from `claimReward`. Updated.